File: apps/jvm/java-console/src/main/java/eu/sirotin/kotunil/app/java/DerivedAbstractTest.java
/*
 * Copyright (c) 2022.  Viktor Sirotin
 *
 *  *  Permission is hereby granted, free of charge, to any person obtaining a copy
 *  * of this software and associated documentation files (the "Software"), to deal
 *  * in the Software without restriction, including without limitation the rights
 *  * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 *  * copies of the Software, and to permit persons to whom the Software is
 *  *  furnished to do so, subject to the following conditions:
 *  *  This Copyright header should still in file if you use it file without changes or
 *  * with changes or if you copy essential part of this file in some new file.
 *  *  *
 *  * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 *  * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 *  * FITNESS FOR A PARTICULAR PURPOSE AND NON-INFRINGEMENT. IN NO EVENT SHALL THE
 *  * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 *  * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 *  * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 *  * THE SOFTWARE.
 *
 */

package eu.sirotin.kotunil.app.java;
import eu.sirotin.kotunil.core.DerivedUnit;
import eu.sirotin.kotunil.core.Expression;
import eu.sirotin.kotunil.derived.Volt;
import org.jetbrains.annotations.NotNull;



class Volt1 extends DerivedUnit {

    public Volt1(@NotNull Number value) {
        super(value, new Volt(1));
    }
}

public class DerivedAbstractTest {

    public void testDerivedAbstract() {
        Expression v1 = new Volt1(24.9);
        Expression v2 = new Volt(24.9);
        Checker.check(v1, v2);
    }

}

File: kotunil/src/commonTest/kotlin/eu/sirotin/kotunil/derived/DerivedAbstractTest.kt
        
package eu.sirotin.kotunil.derived

import eu.sirotin.kotunil.base.A
import eu.sirotin.kotunil.base.kg
import eu.sirotin.kotunil.base.m
import eu.sirotin.kotunil.base.s
import eu.sirotin.kotunil.core.*
import eu.sirotin.kotunil.specialunits.m2
import kotlin.test.assertEquals
import kotlin.test.Test


internal class DerivedAbstractTest {

    private class Volt1(value: Number) : DerivedUnit(value, kg*m2*(s `^` -3) * (A `^` -1))

    @Test
    fun testCoreAndDerived() {

        val x = Newton(12)
        val y = x/m;
        val yc = y.categorySymbols()
        assertEquals("MT-2", yc)
    }

    @Test
    fun testPseudoVolt() {
        assertEquals<Expression>(Volt1(12.6) , Volt(12.6))
    }
}